app/services/oidc_jwt_token_validate.rb

```diff
@@ -0,0 +1,45 @@
+class OidcJwtTokenValidate
+  attr_reader :decoded_token
+  delegate :logger, to: :Rails
+
+  def initialize(jwt_token)
+    @jwt_token = jwt_token
+  end
+
+  def valid_jwt?
+    # OpenSSL#set_key method does not support ruby version < 2.4.0, apparently the JWT gem uses
+    # OpenSSL#set_key method for all ruby version. We must remove this condition once new version
+    # of the JWT(2.2.2) is released.
+    unless OpenSSL::PKey::RSA.new.respond_to?(:set_key)
+      Foreman::Logging.logger('notifications').error "SSO feature is not available for Ruby < 2.4.0"
+      return false
+    end
+    return false if Setting['oidc_algorithm'].nil?
+    JWT.decode(@jwt_token, nil, true,
+      { aud: Setting['oidc_audience'],
+        verify_aud: true,
+        iss: Setting['oidc_issuer'],
+        verify_iss: true,
+        algorithms: [Setting['oidc_algorithm']],
+        jwks: jwks_loader }
+    ).first.present?
+  rescue JWT::DecodeError => e
+    Foreman::Logging.exception('Failed to decode JWT', e)
+    false
+  end
+
+  private
+
+  def jwks_loader
+    response = RestClient::Request.execute(
+      :url => Setting['oidc_jwks_url'],
+      :method => :get,
+      :verify_ssl => true
+    )
+    jwks_keys = JSON.parse(response)['keys']
```

Comment:
`JSON.parse(response)` may not be a hash in which case it can fail. I don't think you're catching that exception.

Reply:
Since you asked the question on IRC:
> For this comment, when can a JSON not be a key-value pair?
Never assume content that comes from the network is in the shape you want. It's always good to think about failure cases. The user might have configured the URL incorrect and it's actually pointing to a different service that does return JSON for example.